### PR TITLE
Streaming support

### PR DESCRIPTION
**Pull request by schmod**
Adds streaming support to mysqlService and mysqlDao

### CODE REVIEW

File: lib/mysqlDao.js [hunk 1]
```diff
@@ -356,6 +357,129 @@ class MysqlDao {
...
+
+        /** @type {MysqlGetFromSqlOptions} */
+        const optionsToUse = _.defaults(options, defaultOptions);
+
+        const defaultStreamOptions = {
+            objectMode: true
+        };
+
+        /** @type {ReadableOptions} **/
+        const streamOptionsToUse = Object.assign(defaultStreamOptions, streamOptions);
```
Comment (cdeveas) [inline lib/mysqlDao.js]: You could use defaults twice and simplify these lines a bit: 
```javascript
options = _.defaults(options, {
    rawResults: false,
    forceReadFromMaster: false
});

streamOptions = _.defaults(streamOptions, {
    objectMode: true
});
```

File: lib/mysqlService.js [hunk 2]
```diff
@@ -231,6 +221,59 @@ class MysqlService {
...
+     * @param pool
+     * @param {String} sql
+     * @param  {ReadableOptions} streamOptions
+     * @return {stream.Readable}
+     */
+    _streamQuery(pool, sql, streamOptions) {
+      this._logSqlQuery(sql);
+
+      const defaults = { objectMode: true };
+      const opts = Object.assign(defaults, streamOptions);
```
Comment (cdeveas) [inline lib/mysqlService.js]: Same comment as above for setting defaults. I usually just do it at the top and overwrite the same options object: 

```javascript
streamOptions = _.defaults(streamOptions, {
    objectMode: true
});
```

File: lib/mysqlService.js [hunk 3]
```diff
@@ -231,6 +221,59 @@ class MysqlService {
...
+
+      const transform = new MysqlService._TransformResult(opts);
+      return queryStream.pipe(transform);
+    }
+
+    static get _TransformResult() {
+      class TransformResult extends TransformStream {
+        _transform(data, encoding, cb) {
+          // convert RowDataPacket to a standard JS object
+          cb(null, Object.assign({}, data));
```
Comment (cdeveas) [inline lib/mysqlService.js]: Looks like you use Object.assign({}, data) twice with that comment. Maybe just roll it into a private function called `_convertToJsObject(row)` which will DRY up and use the function to comment.

File: lib/mysqlDao.js [hunk 4]
```diff
@@ -356,6 +357,129 @@ class MysqlDao {
     }
 
     /**
+     * @param {object} whereClause
+     * @param {MysqlDaoGetAllOptions}
+     * @param {ReadableOptions} [streamOptions]
```
Comment (cdeveas) [inline lib/mysqlDao.js]: Is `ReadableOptions` defined anywhere?

File: lib/mysqlDao.js [hunk 5]
```diff
@@ -356,6 +357,129 @@ class MysqlDao {
...
+    streamFromSqlFromMaster(sql, streamOptions) {
+        return this._streamFromSql(sql, {forceReadFromMaster: true}, streamOptions);
+    }
+
+    /**
+     * @param {string} sql
+     * @param {ReadableOptions} [streamOptions]
+     * @returns {stream.Readable}
+     */
+    streamFromSqlFromMasterRaw(sql, streamOptions) {
```
Comment (cdeveas) [inline lib/mysqlDao.js]: Minor, but adding external api functions for all these different permutations of options may be overkill.

File: test/unit/test-mysqlService.js [hunk 6]
```diff
@@ -292,6 +292,32 @@ describe('MysqlService', function() {
...
+
+                results[1].should.not.have.property('id');
+                results[1].should.have.property('colInt');
+                results[1].colInt.should.eql(3);
+                results[1].should.have.property('colVarchar');
+                results[1].colVarchar.should.eql('three');
+                done();
+              });
+          })
+        })
```
Comment (cdeveas) [inline test/unit/test-mysqlService.js]: Errors can be tricky with streams. If convenient, I'd add a test case that simulates an error on the stream.

Comment (jonathan-fulton) [inline test/unit/test-mysqlService.js]: +1

**Comment by jonathan-fulton** [comment]
@schmod Thanks for submitting this pull request!  Looks like you guys forked your own repo for long term?  J/w

**Comment by schmod** [comment]
Unclear what the long-term future of the fork is going to be.  For now, it's just a place for us to stage PRs, and have a place to publish temporary hotfixes/experiments until we can get a private NPM registry properly set up.

**Comment by schmod** [comment]
(Which is to say, there's no current plan to take the library in a different direction, or make anything other than incremental improvements)

**Comment by jonathan-fulton** [comment]
Ok, sounds good.  Thanks for the update!

File: lib/mysqlDao.js (hunk 4 above)
Comment (cdeveas) [inline lib/mysqlDao.js]: Eh seems like you can leave it if WebStorm figures it out with the typescript library. @typedef is probably overkill

**Pull request merged**